File: Python/19_edit_distance.py
import sys

from collections import namedtuple

Penalties = namedtuple('Penalties', ['insertion', 'deletion', 'substitution'])


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def calc_edit_distance(string_1, string_2, n, m, penalties):
    edit_distance = [[0] * (m+1) for _ in range(n+1)]
    for ix in range(1, n+1):
        edit_distance[ix][0] = ix * penalties.deletion
    for jy in range(1, m+1):
        edit_distance[0][jy] = jy * penalties.insertion

    for ix in range(n):
        for jy in range(m):
            deletion = edit_distance[ix][jy+1] + penalties.deletion
            insertion = edit_distance[ix+1][jy] + penalties.insertion
            match = edit_distance[ix][jy]
            if string_1[ix] != string_2[jy]:
                match += penalties.substitution

            edit_distance[ix+1][jy+1] = min(insertion, deletion, match)
    return edit_distance[-1][-1]


def main():
    reader = sys.stdin
    n, m = convert_to_intlist(next(reader))
    s1 = next(reader).rstrip()
    s2 = next(reader).rstrip()
    penalties = Penalties(*convert_to_intlist(next(reader)))
    dist = calc_edit_distance(s1, s2, n, m, penalties)
    print(dist)


if __name__ == "__main__":
    main()

File: Haskell/19-edit-distance.hs
import Data.Array (Array, listArray, range, (!))

data Penalties = Penalties Int Int Int


convertToIntList :: String -> [Int]
convertToIntList = map read . words


calcEditDistance :: String -> String -> Int -> Int -> Penalties -> Int
calcEditDistance s1 s2 n m penalties =
    let Penalties ip dp sp = penalties
        (a1, a2) = (listArray (1, n) s1, listArray (1, m) s2)
        editDistance :: Int -> Int -> Int
        editDistance ix 0 = ix * dp
        editDistance 0 jy = jy * ip
        editDistance ix jy =
            let matchingScore = if (a1 ! ix == a2 ! jy) then 0 else sp
                deletion = editDistanceValues ! (ix - 1, jy) + dp
                insertion = editDistanceValues ! (ix, jy - 1) + ip
                matching = editDistanceValues ! (ix - 1, jy - 1) + matchingScore
            in minimum [deletion, insertion, matching]
        bounds = ((0, 0), (n, m))
        editDistanceValues = listArray bounds [editDistance ix jy | (ix, jy) <- range bounds]
    in editDistance n m


main :: IO()
main = do
    [n, m] <- fmap convertToIntList getLine
    s1 <- getLine
    s2 <- getLine
    penalties <- fmap ((\[ip, dp, sp] -> Penalties ip dp sp) . convertToIntList) getLine
    let result = calcEditDistance s1 s2 n m penalties
    print result
